mmdet/models/dense_heads/dino_head.py

```diff
@@ -0,0 +1,436 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Dict, List, Tuple
+
+import torch
+from mmengine.structures import InstanceData
+from torch import Tensor
+
+from mmdet.registry import MODELS
+from mmdet.structures import SampleList
+from mmdet.structures.bbox import bbox_cxcywh_to_xyxy, bbox_xyxy_to_cxcywh
+from mmdet.utils import InstanceList, OptInstanceList, reduce_mean
+from ..utils import multi_apply
+from .deformable_detr_head import DeformableDETRHead
+
+
+@MODELS.register_module()
+class DINOHead(DeformableDETRHead):
+    r"""Head of the DINO: DETR with Improved DeNoising Anchor Boxes
+    for End-to-End Object Detection
+
+    Code is modified from the `official github repo
+    <https://github.com/IDEA-Research/DINO>`_.
+
+    More details can be found in the `paper
+    <https://arxiv.org/abs/2203.03605>`_ .
+    """
+
+    def loss(self, hidden_states: Tensor, references: List[Tensor],
+             enc_outputs_class: Tensor, enc_outputs_coord: Tensor,
+             batch_data_samples: SampleList, dn_meta: Dict[str, int]) -> dict:
+        """Perform forward propagation and loss calculation of the detection
+        head on the queries of the upstream network.
+
+        Args:
+            hidden_states (Tensor): Hidden states output from each decoder
+                layer, has shape (num_decoder_layers, num_query_total, bs,
+                dim), where `num_query_total` is the sum of
+                `num_denoising_query` and `num_matching_query`.
+            references (list[Tensor]): List of the reference from the decoder.
+                The first reference is the `init_reference` (initial) and the
+                other num_decoder_layers(6) references are `inter_references`
+                (intermediate). The `init_reference` has shape (bs,
+                num_query_total, 4) and each `inter_reference` has shape
```

Comment:
what is the difference between `num_query_total` and `num_query`?

Reply:
`num_query_total = num_denoising_query + num_matching_query`, i think the `num_query` is more likely to be understood as the `num_matching_query`. Hence, the `total` was added to tell it is the total number of the two kinds of queries, which eliminates ambiguity.